tests/zfs-tests/tests/functional/cli_root/zfs_program/zfs_program_json.ksh

```diff
@@ -0,0 +1,130 @@
+#!/bin/ksh -p
+#
+# CDDL HEADER START
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy is of the CDDL is also available via the Internet
+# at http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2018 Datto Inc.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+#
+# STRATEGY:
+#	1. Compare JSON output formatting for a channel program to template
+#	2. Using bad command line option (-Z) gives correct error output
+#
+
+verify_runnable "both"
+
+function cleanup
+{
+	log_must zfs destroy $TESTDS
+	return 0
+}
+log_onexit cleanup
+
+log_assert "Channel programs output valid JSON"
+
+TESTDS="$TESTPOOL/zcp-json"
+log_must zfs create $TESTDS
+
+TESTZCP="/$TESTDS/zfs_rlist.zcp"
+cat > "$TESTZCP" << EOF
+	succeeded = {}
+	failed = {}
+
+	function list_recursive(root, prop)
+		for child in zfs.list.children(root) do
+			list_recursive(child, prop)
+		end
+		val, src  = zfs.get_prop(root, prop)
+		if (val == nil) then
+			failed[root] = val
+		else
+			succeeded[root] = val
+		end
+	end
+
+	args = ...
+
+	argv = args["argv"]
+
+	list_recursive(argv[1], argv[2])
+
+	results = {}
+	results["succeeded"] = succeeded
+	results["failed"] = failed
+	return results
+EOF
+
+# 1. Compare JSON output formatting for a channel program to template
+typeset -a pos_cmds=("recordsize" "type")
+typeset -a pos_cmds_out=(
+"{
+    \"return\": {
+        \"failed\": {},
+        \"succeeded\": {
+            \"$TESTDS\": 131072
+        }
+    }
+}"
+"{
+    \"return\": {
+        \"failed\": {},
+        \"succeeded\": {
+            \"$TESTDS\": \"filesystem\"
+        }
+    }
+}")
+typeset -i cnt=0
+typeset cmd
+for cmd in ${pos_cmds[@]}; do
+	log_must zfs program $TESTPOOL $TESTZCP $TESTDS $cmd 2>&1
+	log_must zfs program $TESTPOOL -j $TESTZCP $TESTDS $cmd 2>&1
+	OUTPUT=$(zfs program $TESTPOOL -j $TESTZCP $TESTDS $cmd 2>&1 | python -m json.tool)
+	if [ "$OUTPUT" != "${pos_cmds_out[$cnt]}" ]; then
+		echo "Got     :$OUTPUT"
```

Comment:
You might want to use `log_note` in place of `echo` so the output shows up in the logs.

Reply:
Yeah using `log_note` is cleaner. I did find the output though and it doesn't make sense as the strings seem to match yet they don't match.
From http://build.zfsonlinux.org/builders/CentOS%206%20x86_64%20%28TEST%29/builds/2155/steps/shell_8/logs/log:
```
22:58:09.49 Got     :{
22:58:09.49     "return": {
22:58:09.49         "failed": {}, 
22:58:09.49         "succeeded": {
22:58:09.49             "testpool/zcp-json": 131072
22:58:09.49         }
22:58:09.49     }
22:58:09.49 }
22:58:09.49 Expected:{
22:58:09.49     "return": {
22:58:09.49         "failed": {},
22:58:09.49         "succeeded": {
22:58:09.49             "testpool/zcp-json": 131072
22:58:09.49         }
22:58:09.49     }
22:58:09.49 }
22:58:09.49 Unexpected channel program output
```

Only thing I can think of is centos6's json.tool uses tabs instead of spaces...

Reply:
There's a trailing space after the `,` in the "failed": {}" line on CentOS 6.  Trimming the trailing whitespace resolved the test failure for me.

```
       OUTPUT=$(zfs program $TESTPOOL -j $TESTZCP $TESTDS $cmd 2>&1 | python -m json.tool | sed 's/[[:space:]]*$//')
```

Reply:
thanks for finding that, I've updated the PR with your fix, let's see what buildbot thinks...